editor/src/clj/util/http_server.clj

```diff
@@ -13,98 +13,415 @@
+      (log/info :msg "Http server running" :local-url (local-url server)))
+    server))
+
+(defn stop!
+  ^HttpServer
+  ([server] (stop! server 2))
+  ([^HttpServer server delay-seconds]
+   (doto server (.stop delay-seconds))))
+
+(defn router-handler
+  "Create HTTP request handler function from reitit routes
+
+  Routes data could be a map of request methods to handler functions, e.g.:
 
-(defn start! [^HttpServer server]
-  (.start server)
-  (when-not (Boolean/getBoolean "defold.tests")
-    (log/info :msg "Http server running"
-              :local-url (local-url server)))
-  server)
+  {\"/status\" {\"GET\" get-status}
+   \"/resources/{*path}\" {\"GET\" get-resource}
+   \"/command\" {\"GET\" list-commands}}
+   \"/command/{command}\" {\"POST\" invoke-command}
 
-(defn stop! [^HttpServer server]
-  (.stop server 2)
-  server)
+  Use {name} in path to extract a single part of the path, or {*name} to extract
+  the rest of the path until the end. Returned router:
+  - strips or adds trailing slashes if needed to make a match
+  - supports OPTIONS (CORS) requests
+  - automatically supports HEAD requests when GET handler is defined
 
+  Router handler functions receive an HTTP request and return either a response
+  or a CompletableFuture that will contain the response.
+  Request is a map with the following keys:
+    :method         HTTP request method string, e.g. \"GET\"
+    :path           URI path part string, e.g. \"/users/1\"
+    :path-params    a possibly empty map of path extractor pattern (keyword) to
+                    extracted value (string)
+    :query          optional query string, e.g. \"q=1\"
+    :headers        a map of headers, string->(string|string[]); header names
+                    are lower-case
+    :body           request body, InputStream
+  Response is a map with the following keys:
+    :status     optional HTTP response status, integer, default 200
+    :headers    optional response headers map, string->string;
+                header names should be lower-case
+    :body       optional response body, could be nil, string
+                content, or anything that satisfies io/IOFactory
+                (e.g. InputStream, File, Path etc.). If body is AutoCloseable,
+                it will be closed even if it's not written (which might happen
+                when responding to HEAD requests)
+  See also:
+    https://cljdoc.org/d/metosin/reitit-core/0.8.0-alpha1/doc/introduction"
+  [routes]
+  (let [allowed-methods (fn allowed-methods [method->handler]
+                          (let [method-set (-> method->handler keys set (conj "OPTIONS"))
+                                method-set (cond-> method-set (contains? method-set "GET") (conj "HEAD"))]
+                            (string/join ", " (sort method-set))))
+        invoke-handler (fn invoke-handler [handler request match]
+                         (handler (assoc request :path-params (:path-params match))))
```

Comment:
These functions don't appear to use anything from the `routes`, so they could be global?

Reply:
Yes, I'll make them global (in another PR)